base/expr.jl

```diff
@@ -442,3 +445,103 @@ macro generated(f)
         error("invalid syntax; @generated must be used with a function definition")
     end
 end
+
+
+"""
+    @atomic var
+    @atomic order ex
+
+Mark `var` or `ex` as being performed atomically, if `ex` is a supported expression.
+
+See [atomics](#man-atomics) in the manual for more details.
+
+```jldoctest
+julia> mutable struct Atomic{T}; @atomic x::T; end
+
+julia> a = Atomic(1)
+Atomic{Int64}(1)
+
+julia> @atomic :sequentially_consistent a.x = 2 # set field x of a, with sequential consistency
+2
+
+julia> @atomic a.x # fetch field x of a, with sequential consistency
+2
+
+julia> @atomic a.x += 1 # increment field x of a, with sequential consistency
+3
+
+julia> y = 4; @atomic a.x, z = y, a.x # swap field x of a with y and put the old value in z
+(4, 3)
+
+julia> z
+3
+
+julia> y = 5; @atomic a.x, z = y, _ # swap field x of a with y and put the old value in z
+(5, 4)
+
+julia> z
+4
+```
+
+The following forms are also planned, but not yet implemented:
+```julia
+# @atomic a.x = +(a.x, 1) # increment field x of a
+# @atomic a.x = +(_, 1) # increment field x of a
+```
+"""
+macro atomic(ex)
+    if !isa(ex, Symbol) && !is_expr(ex, :(::))
+        return make_atomic(QuoteNode(:sequentially_consistent), ex)
+    end
+    return esc(Expr(:atomic, ex))
+end
+macro atomic(order, ex)
+    return make_atomic(order, ex)
+end
+function make_atomic(@nospecialize(order), @nospecialize(ex))
+    order isa QuoteNode || (order = esc(order))
+    if ex isa Expr
+        if ex.head === :call
+            error("@atomic modifyproperty! syntax not implemented")
+        elseif ex.head === :.
+            l, r = esc(ex.args[1]), esc(ex.args[2])
+            return :(getproperty($l, $r, $order))
```

Comment:
I agree something like `@atomic a.x += x + 1` and `@atomic a.x = (0, 0)` can be handy. But I see `@atomic` as an expert feature so I feel starting from a restrictive API makes sense. Since a line with `@atomic` will be a very delicate part of the code, making this expression as minimal as possible can promote the readability of the code.

Adding to my example abode, it's also not clear to me what `@atomic a.x += 2 * a.x` should do. It's conceivable (= I'm not suggesting it) to lower this to

```julia
op(x, _) = x + 2 * x
modifyproperty!(a, :x, op, nothing, ...)
```

which in turn is compiled to a CAS loop.

Reply:
yeah, that seems a bit too magic to scan for expressions. If you want it, you can write it with anonymous functions:
```julia
@atomic a.x = ((x, _) -> x + 2 * x)(a.x, nothing)
```

Otherwise, you just need to observe a rule that each atomic operation should appear only once in the original expression. Since `a.x` appeared twice, that expression will happen twice. Both atomically, but twice. (one sequentially_consistent, since it pairs with the store, and the other relaxed, since it doesn't have an ordering, either in the original expression nor with a simultaneous store)

Reply:
> Otherwise, you just need to observe a rule that each atomic operation should appear only once in the original expression.

My point is that it sounds better to mechanically verify this at the macro expansion time.

> Since `a.x` appeared twice, that expression will happen twice.

Yeah, but, since this is not the case for `@atomic a.x, z = y, a.x`, I think it's hard to remember and be sure when `a.x` will be pattern-matched and "fused"